# Pangenome Guided Genome Assembly Workflow


## Directory Structure
* `workflow/`: Contains the main Snakefile (Other snakefiles are inside `other_snakefiles/`)
* `config/`: Contains snakemake configuration files for different datasets
* `benchmark/`: Runtime and resource usage statistics for individual Snakemake rules
* `logs/`: Standard output and error logs for each rule execution, useful for debugging
* `results*/`: Output files generated by the pipeline.


## Run the workflow
From the pga_workflow/ directory, run:
```sh
snakemake --use-singularity --singularity-args "-B /private/groups/" --configfile config/config_ontR10.yaml --cores 128 --printshellcmds
```
To resume the workflow from where it left off (e.g., after interruption), use:
```sh
snakemake --use-singularity --singularity-args "-B /private/groups/" --configfile config/config_ontR10.yaml --cores 128 --printshellcmds --rerun-incomplete 
## Sometimes, it asks to unlock and then rerun
snakemake --use-singularity --singularity-args "-B /private/groups/" --configfile config/config_ontR10.yaml --cores 128 --printshellcmds --unlock
```
This will rerun any jobs that were incomplete or failed during the previous run.

## Configuration

The workflow is configured using the `config/config_ontR10.yaml` file. This file contains paths to input files, tool parameters, and other settings. The config file also contains details on the required format of the inputs.

### Key Configuration Parameters
*   `SAMPLE_IDS`: A list of sample IDs to process.
*   `graph_base`: Path to the pangenome graph.
*   `chm13_ref`: Path to the CHM13 reference genome. ([T2T-CHM13v2.0](https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/analysis_set/chm13v2.0.fa.gz) currently being used)
*   `HG002v101_ref`: Path to the HG002 reference genome. ([HG002v1.0.1](https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/hg002v1.1.fasta.gz) currently being used)
*   `MINIMAP`: Parameters for `minimap2`.
*   `HAPLOTYPE_SAMPLING`: Parameters for haplotype sampling.
*   `ANALYSEPAF`, `DISPLAYPAF`, `SHASTA`: Paths to binaries.
*   `VG_ANCHORS`: Specify path to the config file (default: `config/vg_anchors_config.ini`)
*   `TMPDIR`: A temporary directory for intermediate files.
*   `region`: Defines the genomic region to be assembled.


### Workflow Behavior Flags
Additional flags in the config file allow for fine-grained control over the workflow's behavior:
- **`RUN_MODE`**: Controls which workflow to run
  - `"no_positive_control"`: Runs only the sample workflow
  - `"positive_control_only"`: Runs only the positive control workflow. Uses the HG002-included graph for analysis

- **`READ_TYPE`**: Controls which type of reads to use.
  - `"raw"`: Use the original, raw sequencing reads.
  - `"ec"`: Use error-corrected reads generated by a tool like `hifiasm`.

- **`RUN_GBZ_QUERY`** (boolean): If `True`, the workflow will use [gbz-base](https://github.com/jltsiren/gbz-base) `query` to extract subgraph files (`pg.vg`, `gfa`, `dist`, `gaf`). If `False`, it will use `vg chunk`.

- **`USE_FULL_GRAPH`** (boolean): If `True`, anchor generation will be performed using the full graph along with the reads and index from the `gbz` query. **Note:** This will overwrite the `anchors/` directory.

- **`RUN_DEBUGGING`** (boolean): If `True`, the workflow will generate extra output files useful for debugging, such as reliable snarl stats and files for visualization in Bandage.
